client/protocol.go

```diff
@@ -0,0 +1,413 @@
+	fastResCh := make(chan *ProposeResponse)
+	fastErrCh := make(chan error)
+	slowResCh := make(chan *ProposeResponse)
+	slowErrCh := make(chan error)
+
+	go func() {
+		proposeRes, err := c.slowRound(cmd)
+		slowResCh <- proposeRes
+		slowErrCh <- err
+	}()
+
+	go func() {
+		proposeRes, err := c.fastRound(cmd)
+		fastResCh <- proposeRes
+		fastErrCh <- err
+	}()
+
+	if useFastPath {
+		select {
+		case proposeRes := <-fastResCh:
+			err := <-fastErrCh
+			if proposeRes.IsFastRoundSuccess {
+				return proposeRes, err
+			}
+			proposeRes = <-slowResCh
+			err = <-slowErrCh
+			return proposeRes, err
+		case proposeRes := <-slowResCh:
+			err := <-slowErrCh
+			return proposeRes, err
+		}
+	} else {
+		<-fastResCh
+		<-fastErrCh
+		proposeRes := <-slowResCh
+		err := <-slowErrCh
+		return proposeRes, err
+	}
+}
+
+// The fast round of Curp protocol
+// It broadcast the requests to all the curp servers.
+func (c protocolClient) fastRound(cmd *xlineapi.Command) (*ProposeResponse, error) {
+	logrus.Info("Fast round start. Propose ID:", cmd.ProposeId)
+
+	isReceivedLeaderReq := false
+	isReceivedSuccessRes := 0
+	okCnt := 0
+	var clients []curpapi.ProtocolClient
+	var cmdRes xlineapi.CommandResponse
+	var proposeErr xlineapi.ExecuteError
+	proposeCh := make(chan *curpapi.ProposeResponse)
+
+	for _, m := range c.addrs {
+		conn, err := grpc.Dial(m, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(c.timeout.idleTimeout))
+		if err != nil {
+			logrus.Warn(err)
+		} else {
+			protocolClient := curpapi.NewProtocolClient(conn)
+			clients = append(clients[:], protocolClient)
+		}
+	}
+
+	serializeCmd, err := proto.Marshal(cmd)
+	if err != nil {
+		return nil, errors.New("command marshal fail")
+	}
+
+	for _, protocolClient := range clients {
+		client := protocolClient
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), c.timeout.proposeTimeout)
+			defer cancel()
+			res, err := client.Propose(ctx, &curpapi.ProposeRequest{Command: serializeCmd})
+			if err != nil {
+				logrus.Debug(err)
+				proposeCh <- nil
+			} else {
+				proposeCh <- res
+			}
+		}()
+	}
+
+	for i := 0; i < len(clients); i++ {
+		proposeRes := <-proposeCh
+
+		if proposeRes == nil {
+			continue
+		}
+
+		if proposeRes.ExeResult != nil {
+			isReceivedLeaderReq = true
+		}
+
+		switch proposeResp := proposeRes.ExeResult.(type) {
+		case *curpapi.ProposeResponse_Result:
+			switch cmdResult := proposeResp.Result.Result.(type) {
+			case *curpapi.CmdResult_Er:
+				err := proto.Unmarshal(cmdResult.Er, &cmdRes)
+				if err != nil {
+					logrus.Debug(err)
+				} else {
+					isReceivedSuccessRes = 1
+					okCnt++
+				}
+			case *curpapi.CmdResult_Error:
+				err := proto.Unmarshal(cmdResult.Error, &proposeErr)
+				if err != nil {
+					logrus.Debug(err)
+				} else {
+					isReceivedSuccessRes = 2
+					okCnt++
+				}
+			default:
+				logrus.Warn("unknown command result type")
+			}
```

Comment:
This branch means client receive the leader's response.

Reply:
done